presto-spi/src/main/java/com/facebook/presto/spi/memory/ArrayPool.java

```diff
@@ -0,0 +1,345 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.memory;
+
+import com.facebook.presto.spi.PrestoException;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.facebook.presto.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
+import static java.lang.Math.max;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class ArrayPool<T>
+{
+    public abstract static class Allocator<T>
+    {
+        abstract T allocate(int size);
+
+        abstract void initialize(T array);
+
+        abstract int getSize(T array);
+    }
+
+    private static class LongArrayList
+    {
+        private static final int INITIAL_SIZE = 10;
+        private long[] longs = new long[INITIAL_SIZE];
+
+        private int size;
+
+        void add(long value)
+        {
+            if (longs.length <= size) {
+                longs = Arrays.copyOf(longs, 2 * size);
+            }
+            longs[size++] = value;
+        }
+
+        int size()
+        {
+            return size;
+        }
+
+        long get(int i)
+        {
+            if (i >= size) {
+                throw new PrestoException(GENERIC_INTERNAL_ERROR, format("LongArrayList index out of bound %d", i));
+            }
+            return longs[i];
+        }
+
+        void popBack()
+        {
+            size--;
+        }
+
+        // Remove elements from position 0 to index - 1
+        void removeFront(int index)
+        {
+            verify(index >= 0 && index <= size, format("LongArrayList index %d is out of bound", index));
+            size -= index;
+            if (size > index) {
+                System.arraycopy(longs, index, longs, 0, size - index);
+            }
+        }
+    }
+
+    private static class PoolStats
+    {
+        private long totalSize;
+        private long objectCount;
+
+        PoolStats(long totalSize, long objectCount)
+        {
+            this.totalSize = totalSize;
+            this.objectCount = objectCount;
+        }
+    }
+
+    // The fraction of objects that will be sampled to calculate average age during trimming. 2 means 1/2 objects will be sampled.
+    private static final int SAMPLING_FRACTION = 2;
+
+    private final int[] arrayObjectSizes;
+    private ArrayList<T>[] arrayObjects;
```

Comment:
Can we use the interface type where appropriate?  I don't think this will introduce any sort of performance regression.